src/backend/distributed/utils/multi_partitioning_utils.c

```diff
@@ -257,6 +384,146 @@ WorkerFixPartitionConstraintCommandList(Oid relationId, uint64 shardId,
+ */
+static List *
+CreateFixPartitionShardIndexNamesTaskList(Oid relationId)
+{
+	List *taskList = NIL;
+
+	/* enumerate the tasks when putting them to the taskList */
+	int taskId = 1;
+
+	Relation relation = RelationIdGetRelation(relationId);
+	List *indexIdList = RelationGetIndexList(relation);
+
+	/* early exit if the relation does not have any indexes */
+	if (indexIdList == NIL)
+	{
+		RelationClose(relation);
+		return NIL;
+	}
+
+	List *shardIntervalList = LoadShardIntervalList(relationId);
+
+	/* lock metadata before getting placement lists */
+	LockShardListMetadata(shardIntervalList, ShareLock);
+
+	ShardInterval *shardInterval = NULL;
+	foreach_ptr(shardInterval, shardIntervalList)
+	{
+		uint64 shardId = shardInterval->shardId;
+
+		List *queryStringList = WorkerFixPartitionShardIndexNamesCommandList(shardId,
+																			 indexIdList);
+
+		Task *task = CitusMakeNode(Task);
+		task->jobId = INVALID_JOB_ID;
+		task->taskId = taskId++;
+
+		task->taskType = DDL_TASK;
+		SetTaskQueryStringList(task, queryStringList);
+		task->dependentTaskList = NULL;
+		task->replicationModel = REPLICATION_MODEL_INVALID;
+		task->anchorShardId = shardId;
+		task->taskPlacementList = ActiveShardPlacementList(shardId);
+
+		taskList = lappend(taskList, task);
+	}
+
+	RelationClose(relation);
+
+	return taskList;
+}
+
+
+/*
+ * WorkerFixPartitionShardIndexNamesCommandList creates a list of queries that will fix
+ * all child index names of parent indexes on given shard of parent partitioned table.
+ */
+static List *
+WorkerFixPartitionShardIndexNamesCommandList(uint64 shardId, List *indexIdList)
+{
+	List *commandList = NIL;
+	Oid indexId = InvalidOid;
+	foreach_oid(indexId, indexIdList)
+	{
+		/*
+		 * Get the qualified name of the corresponding index of given parent index
+		 * in the parent shard with given shardId
+		 */
+		char *indexName = get_rel_name(indexId);
+		char *shardIndexName = pstrdup(indexName);
+		AppendShardIdToName(&shardIndexName, shardId);
+		Oid schemaId = get_rel_namespace(indexId);
+		char *schemaName = get_namespace_name(schemaId);
+		char *qualifiedShardIndexName = quote_qualified_identifier(schemaName,
+																   shardIndexName);
+
+		/*
+		 * Get the list of all partition indexes that are children of current
+		 * index on parent
+		 */
+		List *partitionIndexIds = find_inheritance_children(indexId,
+															ShareRowExclusiveLock);
+		Oid partitionIndexId = InvalidOid;
+		foreach_oid(partitionIndexId, partitionIndexIds)
+		{
+			/* get info for this partition relation of this index*/
+			char *partitionIndexName = get_rel_name(partitionIndexId);
+			Oid partitionId = IndexGetRelation(partitionIndexId, false);
+			char *partitionName = get_rel_name(partitionId);
+			char *partitionSchemaName = get_namespace_name(get_rel_namespace(
+															   partitionId));
+			List *shardIntervalList = LoadShardIntervalList(partitionId);
+
+			/* lock metadata before getting placement lists */
```

Comment:
It is not typical to acquire these locks so deep inside a function named `WorkerXXX`.

Can we move locking these to `CreateFixPartitionShardIndexNamesTaskList` and maybe follow similar approach as `InterShardDDLTaskList` on how to acquire these locks?

Reply:
I couldn't understand how I can follow the approach from `InterShardDDLTaskList`
Can I do the following in `CreateFixPartitionShardIndexNamesTaskList`
```
List *parentShardIntervalList = LoadShardIntervalList(parentRelationId);
List *partitionList = PartitionList(parentRelationId);
/* lock metadata before getting placement lists */
LockShardListMetadata(parentShardIntervalList, ShareLock);
Oid partitionId = InvalidOid;
foreach(partitionId, partitionList)
{
	List *partitionShardIntervalList = LoadShardIntervalList(partitionId);
	LockShardListMetadata(partitionShardIntervalList, ShareLock);
}
```